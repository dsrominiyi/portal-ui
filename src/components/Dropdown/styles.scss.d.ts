export const dropdown: string;
export const isInvalid: string;
