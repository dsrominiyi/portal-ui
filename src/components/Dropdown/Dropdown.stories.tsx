import React from 'react';
import { storiesOf } from '@storybook/react';
import Dropdown, { PrimitiveOption, DropdownOption, Props } from '.';

const stories = storiesOf('Dropdown', module);

const onChangeDefault = (value: PrimitiveOption) => alert(value);

export const render = (
  options: DropdownOption[],
  selected?: string,
  onChange?: Props['onChange'],
  isInvalid?: boolean
) => (props?: any) => (
  <Dropdown
    key={props && props.key}
    options={options}
    selected={selected}
    onChange={onChange || onChangeDefault}
    isInvalid={isInvalid}
  />
);

export const dropdownTestData = [
  { value: 'value1', text: '1' },
  { value: 'value2', text: '2' },
  { value: 'value3', text: '3' },
  { value: 'value4', text: '4' }
];

const primitiveOptionList = dropdownTestData.map(option => option.text);

export const storyRenders = {
  withoutSelectedValue: render(dropdownTestData),
  withSelectedValue: render(dropdownTestData, 'key2'),
  withPrimitiveOptions: render(primitiveOptionList),
  invalid: render(primitiveOptionList, undefined, undefined, true)
};

stories.add('Without pre-selected value', storyRenders.withoutSelectedValue);
stories.add('With pre-selected value', storyRenders.withSelectedValue);
stories.add('Invalid', storyRenders.invalid);
