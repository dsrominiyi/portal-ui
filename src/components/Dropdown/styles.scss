@import '../../style/vars';

.dropdown {
  display: inline-block;
  height: 40px;
  min-width: 140px;
  padding: 0 40px 0 20px;
  border: 1px solid $brand-off-white;
  border-radius: 0;
  outline: none;
  color: $brand-charcoal;
  background-color: $brand-white;
  background-image: url("data:image/svg+xml;charset=UTF-8,%3Csvg%20width%3D%2224%22%20height%3D%2216%22%20viewBox%3D%220%200%2024%2016%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%0A%20%20%20%20%3Cpolygon%20fill%3D%22%23666%22%20points%3D%2212%201%209%206%2015%206%22%20%2F%3E%0A%20%20%20%20%3Cpolygon%20fill%3D%22%23666%22%20points%3D%2212%2013%209%208%2015%208%22%20%2F%3E%0A%3C%2Fsvg%3E%0A");
  background-repeat: no-repeat;
  background-position: 100% 61%;
  font-size: 1rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  cursor: pointer;
  transition: border 0.2s;

  &.is-invalid {
    border: 1px solid $brand-danger;

    &:focus {
      border: 1px solid $brand-danger;
    }
  }

  &:focus {
    border: 1px solid $brand-primary;
  }
}
