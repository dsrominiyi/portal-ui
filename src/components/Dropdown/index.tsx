import React, { FunctionComponent } from 'react';
import classnames from 'classnames/bind';

import formStyles from '../../forms/styles.scss';
import styles from './styles.scss';

export type PrimitiveOption = string | number;

type ObjectOption = {
  text: string;
  value: PrimitiveOption;
};

export type DropdownOption = ObjectOption | PrimitiveOption;

export interface Props {
  label?: string;
  options: DropdownOption[];
  selected?: PrimitiveOption;
  placeholder?: string;
  isInvalid?: boolean;
  attributes?: { [attr: string]: string };
  onChange(value: PrimitiveOption): void;
}

const cx = classnames.bind(styles);

const isPrimitiveOption = (option: DropdownOption): option is PrimitiveOption => {
  return typeof option === 'string' || typeof option === 'number';
};

export const Dropdown: FunctionComponent<Props> = ({
  label,
  options,
  selected,
  placeholder,
  isInvalid,
  attributes = {},
  onChange
}: Props) => {
  const optionValues = options.map((option: DropdownOption) => {
    const value = isPrimitiveOption(option) ? option : option.value;
    const text = isPrimitiveOption(option) ? option : option.text;

    return (
      <option key={value} value={value}>
        {text}
      </option>
    );
  });

  return (
    <div>
      {label && <div className={formStyles.label}>{label}</div>}
      <select
        className={cx({ dropdown: true, isInvalid })}
        onChange={event => onChange(event.target.value)}
        value={selected || ''}
        {...attributes}
      >
        <option key={0} value="" disabled>
          {placeholder || '- Select -'}
        </option>
        {optionValues}
      </select>
    </div>
  );
};

Dropdown.displayName = 'Dropdown';
export default Dropdown;
