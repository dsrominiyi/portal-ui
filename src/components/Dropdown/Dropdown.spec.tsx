import { shallow } from 'enzyme';

import { storyRenders, dropdownTestData, render } from './Dropdown.stories';
import testStorySnapshots from '../../helpers/test/testStorySnapshots';

describe('Dropdown', () => {
  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders);
  });

  test('onChange is called with the newly selected value', () => {
    const onChange = jest.fn();
    const event = { target: { value: 'testVal' } };
    const dropdown = shallow(render(dropdownTestData, undefined, onChange)());
    dropdown.find('select').simulate('change', event);

    expect(onChange).toHaveBeenCalledWith(event.target.value);
  });
});
