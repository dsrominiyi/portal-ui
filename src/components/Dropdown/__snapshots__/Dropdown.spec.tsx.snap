// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component story renders match snapshots: invalid 1`] = `
<div>
  <select
    className="dropdown isInvalid"
    onChange={[Function]}
    value=""
  >
    <option
      disabled={true}
      key="0"
      value=""
    >
      - Select -
    </option>
    <option
      key="1"
      value="1"
    >
      1
    </option>
    <option
      key="2"
      value="2"
    >
      2
    </option>
    <option
      key="3"
      value="3"
    >
      3
    </option>
    <option
      key="4"
      value="4"
    >
      4
    </option>
  </select>
</div>
`;

exports[`Dropdown component story renders match snapshots: withPrimitiveOptions 1`] = `
<div>
  <select
    className="dropdown"
    onChange={[Function]}
    value=""
  >
    <option
      disabled={true}
      key="0"
      value=""
    >
      - Select -
    </option>
    <option
      key="1"
      value="1"
    >
      1
    </option>
    <option
      key="2"
      value="2"
    >
      2
    </option>
    <option
      key="3"
      value="3"
    >
      3
    </option>
    <option
      key="4"
      value="4"
    >
      4
    </option>
  </select>
</div>
`;

exports[`Dropdown component story renders match snapshots: withSelectedValue 1`] = `
<div>
  <select
    className="dropdown"
    onChange={[Function]}
    value="key2"
  >
    <option
      disabled={true}
      key="0"
      value=""
    >
      - Select -
    </option>
    <option
      key="value1"
      value="value1"
    >
      1
    </option>
    <option
      key="value2"
      value="value2"
    >
      2
    </option>
    <option
      key="value3"
      value="value3"
    >
      3
    </option>
    <option
      key="value4"
      value="value4"
    >
      4
    </option>
  </select>
</div>
`;

exports[`Dropdown component story renders match snapshots: withoutSelectedValue 1`] = `
<div>
  <select
    className="dropdown"
    onChange={[Function]}
    value=""
  >
    <option
      disabled={true}
      key="0"
      value=""
    >
      - Select -
    </option>
    <option
      key="value1"
      value="value1"
    >
      1
    </option>
    <option
      key="value2"
      value="value2"
    >
      2
    </option>
    <option
      key="value3"
      value="value3"
    >
      3
    </option>
    <option
      key="value4"
      value="value4"
    >
      4
    </option>
  </select>
</div>
`;
